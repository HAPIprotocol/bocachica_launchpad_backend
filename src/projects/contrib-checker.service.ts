import {
  Inject,
  Injectable,
  Logger,
  OnModuleDestroy,
  OnModuleInit,
} from '@nestjs/common';
import { CronExpression, SchedulerRegistry } from '@nestjs/schedule';
import * as Queue from 'bee-queue';
import { CronJob } from 'cron';

import { Web3Connection, WEB3_CONNECTION } from '../web3/web3.module';
import { flobj } from '../common/string';
import { QUEUE_REDIS_URL } from '../config';
import { ProjectsService } from './projects.service';
import { getProcessType, ProcessType } from '../cluster';

export interface ContribCheckerJob {
  roundId: number;
}

const CONTRIB_CHECKER_JOB_CONCURRENCY = 10;

@Injectable()
export class ContribCheckerService implements OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(ContribCheckerService.name);
  public readonly queue: Queue<ContribCheckerJob>;
  private roundSubs: number[] = [];

  constructor(
    private readonly projectsService: ProjectsService,
    @Inject(WEB3_CONNECTION)
    private readonly web3: Web3Connection,
    private readonly schedulerRegistry: SchedulerRegistry,
  ) {
    if (getProcessType() === ProcessType.Web) {
      const queue = new Queue(ContribCheckerService.name, {
        redis: QUEUE_REDIS_URL,
        isWorker: false,
      });

      this.queue = queue;
    } else {
      const queue = new Queue(ContribCheckerService.name, {
        redis: QUEUE_REDIS_URL,
      });

      queue.process(CONTRIB_CHECKER_JOB_CONCURRENCY, async (job) => {
        const { roundId } = job.data;

        const round = await this.projectsService.getRoundInfo(roundId);

        if (!round) {
          throw new Error(`Round not found: ${roundId}`);
        }

        await this.projectsService.fetchContributions(round);
      });

      queue.on('ready', () => {
        this.logger.log(`Queue is ready`);
      });

      queue.on('error', (err) => {
        this.logger.error(err.message);
      });

      queue.on('job succeeded', (jobId, result) => {
        this.logger.debug(`Job succeeded id=${jobId} total=${result}`);
      });

      queue.on('job retrying', (jobId, error) => {
        this.logger.debug(
          `Job retrying id=${jobId} err=${JSON.stringify(error.message)}`,
        );
      });

      queue.on('job failed', (jobId, error) => {
        this.logger.error(
          `Job failed id=${jobId} err=${JSON.stringify(error.message)}`,
        );
      });

      queue.on('job progress', (jobId, progress) => {
        this.logger.debug(`Job progress id=${jobId} progress=${progress}`);
      });

      this.queue = queue;

      {
        this.logger.log(`Starting cron job checkRoundContributions`);

        const job = new CronJob(CronExpression.EVERY_10_MINUTES, async () => {
          await this.checkRoundContributions();
        });

        this.schedulerRegistry.addCronJob('checkRoundContributions', job);
        job.start();
      }

      {
        this.logger.log(`Starting cron job subscribeForRounds`);

        const job = new CronJob(CronExpression.EVERY_10_MINUTES, async () => {
          await this.subscribeForRounds();
        });

        this.schedulerRegistry.addCronJob('subscribeForRounds', job);
        job.start();
      }
    }
  }

  async checkRoundContributions() {
    const rounds = await this.projectsService.getActiveRounds();
    this.logger.log(
      `Checking round contributions ${flobj({ count: rounds.length })}`,
    );
    for (const round of rounds) {
      this.logger.verbose(
        `Checking round contributions ${flobj({
          roundId: round.id,
          address: round.address,
        })}`,
      );
      await this.updateRoundContribution(round.id);
    }
  }

  async subscribeForRounds() {
    this.logger.debug(
      `Cleaning round account subscriptions ${flobj({
        count: this.roundSubs.length,
      })}`,
    );
    for (const subId of this.roundSubs) {
      try {
        await this.web3.removeAccountChangeListener(subId);
      } catch (_) {
        this.logger.warn(
          `Can't remove account change listener ${flobj({ subId })}`,
        );
      }
    }
    this.roundSubs = [];

    const rounds = await this.projectsService.getActiveRounds();
    for (const round of rounds) {
      const tokenAddress = await round.tokenAddress();

      this.logger.log(
        `Subscribing to round address updated ${flobj({
          roundId: round.id,
          address: round.address,
          tokenAddress: tokenAddress.toString(),
        })}`,
      );

      const subId = this.web3.onAccountChange(tokenAddress, async () => {
        this.logger.log(
          `Round address updated ${flobj({
            roundId: round.id,
            address: round.address,
            tokenAddress: tokenAddress.toString(),
          })}`,
        );
        await this.updateRoundContribution(round.id);
      });
      this.roundSubs.push(subId);
    }
  }

  async onModuleInit() {
    await this.queue.ready();
    if (getProcessType() === ProcessType.Worker) {
      await this.checkRoundContributions();
      await this.subscribeForRounds();
    }
  }

  async onModuleDestroy() {
    await this.queue.close();
  }

  async updateRoundContribution(roundId: number): Promise<string> {
    return new Promise(async (resolve, reject) => {
      const job = await this.queue
        .createJob({ roundId })
        .timeout(60000)
        .retries(3)
        .backoff('exponential', 1000)
        .save();

      job.on('succeeded', (result) => {
        resolve(result);
      });

      job.on('failed', (err) => {
        reject(err);
      });

      this.logger.debug(
        `Job started ${flobj({ id: job.id, roundId: job.data.roundId })}`,
      );
    });
  }
}
