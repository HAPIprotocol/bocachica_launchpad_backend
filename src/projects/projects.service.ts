import { Inject, Injectable, Logger, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Not, Repository } from 'typeorm';

import { Web3Connection, WEB3_CONNECTION } from '../web3/web3.module';
import { DEFAULT_ITEMS_PER_PAGE } from '../config';
import { ProjectContribution } from './entities/project-contribution.entity';
import { ProjectWithCurrentRound } from './dto/find-all-projects.dto';
import { ProjectPartner } from './entities/project-partner.entity';
import {
  ProjectRound,
  ProjectRoundStatus,
} from './entities/project-round.entity';
import { Project } from './entities/project.entity';
import { SolanabeachService } from '../solanabeach/solanabeach.service';
import { flobj } from '../common/string';

@Injectable()
export class ProjectsService {
  private readonly logger = new Logger(ProjectsService.name);

  constructor(
    @InjectRepository(Project) private projectRepo: Repository<Project>,
<<<<<<< HEAD
    @InjectRepository(ProjectContribution)
    private contribRepo: Repository<ProjectContribution>,
    @InjectRepository(ProjectRound) private roundRepo: Repository<ProjectRound>,
    @InjectRepository(ProjectPartner)
    private partnerRepo: Repository<ProjectPartner>,
    @Inject(WEB3_CONNECTION)
    private readonly web3: Web3Connection,
    private readonly solanabeach: SolanabeachService,
=======
    @InjectRepository(ProjectRound) private roundRepo: Repository<ProjectRound>,
    @InjectRepository(ProjectPartner)
    private partnerRepo: Repository<ProjectPartner>,
>>>>>>> ac2163a433fff375279b6ef94418be2cb225ad00
  ) {}

  async findAll(skip = 0, take = DEFAULT_ITEMS_PER_PAGE) {
    const [list, total] = await this.projectRepo.findAndCount({
      order: { id: 'ASC' },
      skip: skip > 0 ? skip : 0,
      take,
    });

    const promises: Promise<void>[] = [];
    for (const project of list as ProjectWithCurrentRound[]) {
      promises.push(
        this.roundRepo
          .findOne({
            where: {
              projectId: project.id,
              status: Not(ProjectRoundStatus.Finished),
            },
            order: { startDate: 'ASC' },
          })
          .then((round) => {
            project.currentRound = round;
          }),
      );
    }
    await Promise.all(promises);

    return { list, total };
  }

  async findOne(id: number) {
    return this.projectRepo
      .findOneOrFail({
        where: { id },
        relations: ['rounds', 'partners'],
      })
      .catch(() => {
        throw new NotFoundException();
      });
  }

  async getContrib(publicKey: string, roundId: number) {
    const { amount } = await this.contribRepo
      .createQueryBuilder()
      .select('SUM(amount)', 'amount')
      .where({ publicKey, roundId })
      .getRawOne();
    return amount;
  }

  async getRoundInfo(roundId: number) {
    return this.roundRepo.findOne({
      where: { roundId },
      relations: ['project'],
    });
  }

  async fetchContributions(round: ProjectRound) {
    const latestContrib = await this.contribRepo.findOne({
      where: { roundId: round.id },
      order: { blocknumber: 'DESC' },
    });

    const offset = 100;
    let limit = 0;

    while (true) {
      const transfers = await this.solanabeach.getTokenTransfers(
        round.address,
        limit,
        offset,
      );

      // Historic data?
      if (
        transfers[0] &&
        transfers[0].blocknumber >= latestContrib.blocknumber
      ) {
        break;
      }

      for (const transfer of transfers) {
        const publicKey = transfer.source.address;
        const txHash = transfer.txhash;
        const logobj = {
          publicKey,
          txHash,
          amount: transfer.amount,
        };

        if (!transfer.valid) {
          this.logger.debug(`Invalid contrib ${flobj(logobj)}`);
          continue;
        }

        const existing = await this.contribRepo.findOne({
          where: { publicKey, txHash },
        });
        if (existing) {
          this.logger.warn(`Duplicate contrib ${flobj(logobj)}`);
          continue;
        }

        const contrib = new ProjectContribution();
        contrib.roundId = round.id;
        contrib.publicKey = transfer.source.address;
        contrib.blocknumber = transfer.blocknumber;
        contrib.timestamp = new Date(transfer.timestamp.absolute);
        contrib.amount = transfer.amount.toString();
        contrib.txHash = transfer.txhash;
        await this.contribRepo.save(contrib);
        this.logger.log(`Saved contrib ${flobj(logobj)}`);
      }

      // Last page?
      if (transfers.length < limit) {
        break;
      }

      limit += offset;
    }
  }
}
