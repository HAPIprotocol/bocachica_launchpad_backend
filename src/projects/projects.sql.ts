export const collectedAmountSql = `
  select
    coalesce(sum(amount), 0) as amount
  from
    (
    select
      "roundId",
      "publicKey",
      "amount",
      case
        when "solPowerCheck" = true and "solPowerScaling" = true then "solPower" between "minAmount" * "solPowerRate" and "maxAmount" * "solPowerRate"
        when "solPowerCheck" = true then "solPower" > 0
        else true
      end as "allowed"
    from
      (
      select
        pc."roundId",
        pc."publicKey",
        sum(pc."amount") as "amount",
        min(ua."solPowerAmount") as "solPower"
      from
        "project_contribution" pc
      join "user_account" ua on
        pc."publicKey" = ua."publicKey"
      group by
        pc."roundId",
        pc."publicKey") contrib
    join (
      select
        id,
        "minAmount",
        "maxAmount",
        "solPowerRate",
        "solPowerScaling"
      from
        "project_round") pr on
      contrib."roundId" = pr."id") t
  where
    "roundId" = $1 and "allowed" = true`;
