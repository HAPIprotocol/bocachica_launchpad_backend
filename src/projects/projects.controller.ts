import {
  Controller,
  DefaultValuePipe,
  Get,
  Param,
  ParseIntPipe,
  Query,
} from '@nestjs/common';
import {
  ApiNotFoundResponse,
  ApiOkResponse,
  ApiQuery,
  ApiTags,
} from '@nestjs/swagger';

import { FindAllProjectsResultDto } from './dto/find-all-projects.dto';
import { FindAllRoundsResultDto } from './dto/find-all-rounds.dto';
import { FindOneProjectResultDto } from './dto/find-one-project.dto';
import { FindOneRoundResultDto } from './dto/find-one-round.dto';
import { GetRoundContributionDto } from './dto/get-round-contribution.dto';
import {
  ProjectRoundAccessType,
  ProjectRoundStatus,
} from './entities/project-round.entity';
import { ProjectsService } from './projects.service';

@Controller('projects')
@ApiTags('Projects')
export class ProjectsController {
  constructor(private readonly projectsService: ProjectsService) {}

  @Get()
  @ApiOkResponse({
    description: 'Return list of projects',
    type: FindAllProjectsResultDto,
  })
  @ApiQuery({
    name: 'skip',
    type: Number,
    required: false,
  })
  findAll(
    @Query('skip', new DefaultValuePipe(0), ParseIntPipe) skip?: number,
  ): Promise<FindAllProjectsResultDto> {
    return this.projectsService.findAll(skip);
  }

  @Get('roundContribution')
  @ApiOkResponse({
    description: 'Get round contribution for public key',
    type: GetRoundContributionDto,
  })
  async getRoundContribution(
    @Query('publicKey') publicKey: string,
    @Query('roundId', ParseIntPipe) roundId: number,
  ): Promise<GetRoundContributionDto> {
    const amount = await this.projectsService.getContrib(publicKey, roundId);
    return { amount };
  }

  @Get('rounds')
  @ApiOkResponse({
    description: 'Get round list',
    type: FindAllRoundsResultDto,
  })
  @ApiQuery({
    name: 'skip',
    type: Number,
    required: false,
  })
  @ApiQuery({
    name: 'take',
    type: Number,
    required: false,
  })
  @ApiQuery({
    name: 'accessType',
    type: Number,
    required: false,
    enum: ProjectRoundAccessType,
  })
  @ApiQuery({
    name: 'status',
    type: Number,
    required: false,
    enum: ProjectRoundStatus,
  })
  @ApiQuery({
    name: 'publicKey',
    type: String,
    required: false,
  })
  async findAllRounds(
    @Query('skip', new DefaultValuePipe(0), ParseIntPipe) skip: number,
    @Query('take', new DefaultValuePipe(3), ParseIntPipe) take: number,
    @Query('accessType') accessType?: ProjectRoundAccessType,
    @Query('status') status?: ProjectRoundStatus,
    @Query('publicKey') publicKey?: string,
  ): Promise<FindAllRoundsResultDto> {
    return this.projectsService.findAllRounds({
      skip,
      take,
      accessType,
      status,
      publicKey,
    });
  }

  @Get('round/:id')
  @ApiOkResponse({
    description: 'Return project round data',
    type: FindOneRoundResultDto,
  })
  @ApiNotFoundResponse({ status: 404, description: 'Not found' })
  async findOneRound(@Param('id') id: string): Promise<FindOneRoundResultDto> {
    const item = await this.projectsService.findOneRound(+id);
    return { item };
  }

  @Get(':id')
  @ApiOkResponse({
    description: 'Return record data',
    type: FindOneProjectResultDto,
  })
  @ApiNotFoundResponse({ status: 404, description: 'Not found' })
  async findOne(@Param('id') id: string): Promise<FindOneProjectResultDto> {
    const item = await this.projectsService.findOne(+id);
    return { item };
  }
}
