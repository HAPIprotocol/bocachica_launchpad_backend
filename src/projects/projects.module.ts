import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { Web3Module } from '../web3/web3.module';
import { SolanabeachModule } from '../solanabeach/solanabeach.module';
import { ProjectsController } from './projects.controller';
import { ProjectsService } from './projects.service';
import { Project } from './entities/project.entity';
import { ProjectPartner } from './entities/project-partner.entity';
import { ProjectRound } from './entities/project-round.entity';
import { ProjectContribution } from './entities/project-contribution.entity';
import { ContribCheckerService } from './contrib-checker.service';
import { TicketsModule } from '../tickets/tickets.module';
import { TotalSupplyCheckerService } from './total-supply-checker.service';

@Module({
  imports: [
    TypeOrmModule.forFeature([
      Project,
      ProjectPartner,
      ProjectRound,
      ProjectContribution,
    ]),
    Web3Module,
    SolanabeachModule,
    TicketsModule,
  ],
  controllers: [ProjectsController],
  providers: [
    ProjectsService,
    ContribCheckerService,
    TotalSupplyCheckerService,
  ],
})
export class ProjectsModule {}
