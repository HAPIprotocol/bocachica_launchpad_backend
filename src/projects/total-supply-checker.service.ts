import {
  Inject,
  Injectable,
  Logger,
  OnModuleDestroy,
  OnModuleInit,
} from '@nestjs/common';
import { PublicKey } from '@solana/web3.js';
import { CronExpression, SchedulerRegistry } from '@nestjs/schedule';
import * as Queue from 'bee-queue';
import * as BN from 'bn.js';
import { CronJob } from 'cron';

import { Web3Connection, WEB3_CONNECTION } from '../web3/web3.module';
import { getProcessType, ProcessType } from '../cluster';
import {
  QUEUE_REDIS_URL,
  TOTAL_SUPPLY_CHECKER_JOB_CONCURRENCY,
} from '../config';
import { ProjectsService } from './projects.service';
import { flobj } from 'src/common/string';

export interface TotalSupplyCheckerJob {
  projectId: number;
}

@Injectable()
export class TotalSupplyCheckerService
  implements OnModuleInit, OnModuleDestroy
{
  private readonly logger = new Logger(TotalSupplyCheckerService.name);

  public readonly queue: Queue<TotalSupplyCheckerJob>;

  constructor(
    private readonly projectsService: ProjectsService,
    private readonly schedulerRegistry: SchedulerRegistry,
    @Inject(WEB3_CONNECTION)
    private readonly web3: Web3Connection,
  ) {
    if (getProcessType() === ProcessType.Web) {
      const queue = new Queue(TotalSupplyCheckerService.name, {
        redis: QUEUE_REDIS_URL,
        isWorker: false,
      });

      this.queue = queue;
    } else {
      const queue = new Queue(TotalSupplyCheckerService.name, {
        redis: QUEUE_REDIS_URL,
      });

      queue.process(TOTAL_SUPPLY_CHECKER_JOB_CONCURRENCY, async (job) => {
        const projectId = job.data.projectId;

        const project = await this.projectsService.findOne(projectId);

        const {
          value: { amount },
        } = await this.web3.getTokenSupply(
          new PublicKey(project.smartContractAddress),
        );
        await this.projectsService.updateTotalSupply(projectId, amount);

        return new BN(amount).toString();
      });

      queue.on('ready', () => {
        this.logger.log(`Queue is ready`);
      });

      queue.on('error', (err) => {
        this.logger.error(err.message);
      });

      queue.on('job succeeded', (jobId, result) => {
        this.logger.debug(`Job succeeded id=${jobId} totalSupply=${result}`);
      });

      queue.on('job retrying', (jobId, error) => {
        this.logger.debug(
          `Job retrying id=${jobId} err=${JSON.stringify(error.message)}`,
        );
      });

      queue.on('job failed', (jobId, error) => {
        this.logger.error(
          `Job failed id=${jobId} err=${JSON.stringify(error.message)}`,
        );
      });

      queue.on('job progress', (jobId, progress) => {
        this.logger.debug(`Job progress id=${jobId} progress=${progress}`);
      });

      this.queue = queue;
    }
  }

  async onModuleInit() {
    await this.queue.ready();
    if (getProcessType() === ProcessType.Worker) {
      this.logger.log(`Starting cron job updateProjectTotalSupplies`);

      const job = new CronJob(CronExpression.EVERY_HOUR, async () => {
        await this.updateProjectTotalSupplies();
      });

      await this.updateProjectTotalSupplies();

      this.schedulerRegistry.addCronJob('updateProjectTotalSupplies', job);
      job.start();
    }
  }

  async onModuleDestroy() {
    await this.queue.close();
  }

  async updateProjectTotalSupplies() {
    const projects = await this.projectsService.projectsToUpdateTotalSupply();
    this.logger.log('Started updating project total supplies');

    for (const project of projects) {
      const totalSupply = await this.getTotalSupply(project.id);
      await this.projectsService.updateTotalSupply(project.id, totalSupply);
      this.logger.verbose(
        `Updated total supply ${flobj({ projectId: project.id, totalSupply })}`,
      );
    }

    this.logger.log('Finished updating project total supplies');
  }

  async getTotalSupply(projectId: number): Promise<string> {
    return new Promise(async (resolve, reject) => {
      const job = await this.queue
        .createJob({ projectId })
        .timeout(1000)
        .retries(3)
        .backoff('exponential', 1000)
        .save();

      job.on('succeeded', (result) => {
        resolve(result);
      });

      job.on('failed', (err) => {
        reject(err);
      });

      this.logger.debug(
        `Job started id=${job.id} projectId=${job.data.projectId}`,
      );
    });
  }
}
