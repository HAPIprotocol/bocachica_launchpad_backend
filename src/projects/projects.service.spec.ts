import { Test, TestingModule } from '@nestjs/testing';

import {
  IMockEntityProvider,
  mockEntityProvider,
} from '../../test/util/typeorm';
import { Project } from './entities/project.entity';
import { ProjectsService } from './projects.service';

describe('ProjectsService', () => {
  let service: ProjectsService;
  let tableProvider: IMockEntityProvider<Project>;

  beforeEach(async () => {
    tableProvider = mockEntityProvider(Project);

    const module: TestingModule = await Test.createTestingModule({
      providers: [ProjectsService, tableProvider],
    }).compile();

    service = module.get<ProjectsService>(ProjectsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
});
