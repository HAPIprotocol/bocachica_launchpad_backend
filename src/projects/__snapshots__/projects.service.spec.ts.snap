// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsService findAll should find all - basic 1`] = `
Object {
  "list": Array [
    Project {
      "currentRound": ProjectRound {
        "collectedAmount": "0",
        "decimals": 9,
        "id": 1,
        "maxAmount": "0",
        "minAmount": "0",
        "projectId": 1,
        "targetAmount": "0",
      },
      "id": 1,
      "title": "top project",
    },
  ],
  "total": 1,
}
`;

exports[`ProjectsService findAll should find all - empty 1`] = `
Object {
  "list": Array [],
  "total": 0,
}
`;

exports[`ProjectsService findOne should find one 1`] = `
Object {
  "id": 1,
}
`;

exports[`ProjectsService getContrib should get round contribution 1`] = `10`;

exports[`ProjectsService getRoundInfo should get round info 1`] = `
ProjectRound {
  "collectedAmount": "0",
  "decimals": 9,
  "id": 10,
  "maxAmount": "0",
  "minAmount": "0",
  "targetAmount": "0",
}
`;
